haskell-modules/generic-builder.nix: work around libc++abi issue

reckenrode opened the pull request:
## Description of changes

Fixes failing build on staging-next #263535 plus whatever else was broken in nixpkgs-review.

* haskellPackages.inline-c-cpp: https://hydra.nixos.org/build/239811177

This is a workaround until cc-wrapper is fixed to link libc++abi even when invoked as just `clang`. Other approaches were investigated (modifying `haskell-modules/generic-builder.nix`), but they would result in mass rebuilds.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

cdepillabout commented:
> I tried an approach that modified haskell-modules/generic-builder.nix, but that’s not really the right place, and it resulted in tons of rebuilds

I haven't been following these clang changes.  Could you be more specific about where would be the right place to make these fixes?

For instance, if you don't fix `haskell-modules/generic-builder.nix`, will there be lots of downstream breakages in end-user's Haskell projects?

> Once cc-wrapper is fixed so clang includes libc++abi automatically even when it’s invoked as just clang, these commits can be reverted.

Is there an issue/PR  you could link to in the code that made it easier to track when this happens?  (Maybe it is https://github.com/NixOS/nixpkgs/issues/166205?)

Also, if these commits are meant to just all be reverted soon, it might make it easier for us if all of the changes in this PR were included in a single commit.

reckenrode commented:
> I haven't been following these clang changes. Could you be more specific about where would be the right place to make these fixes?

The right place is cc-wrapper. This problem affects any program that invokes `clang` to build C++ code. See also: #216047 and #150655.

My primary concern with modifying `haskell-modules/generic-builder.nix` until cc-wrapper is fixed is the number of rebuilds it will cause.

> For instance, if you don't fix `haskell-modules/generic-builder.nix`, will there be lots of downstream breakages in end-user's Haskell projects?

Depending on which modules are used, potentially.

Should I go ahead with modifying `haskell-modules/generic-builder.nix` even if it’s going to result in rebuilding effectively everything? I can target the change so it doesn’t affect Linux, but there’s no way of avoiding mass rebuilds on Darwin.
 
> > Once cc-wrapper is fixed so clang includes libc++abi automatically even when it’s invoked as just clang, these commits can be reverted.
> 
> Is there an issue/PR you could link to in the code that made it easier to track when this happens? (Maybe it is #166205?)

I pinged the [Nixpkgs/NixOS Contributions](https://matrix.to/#/#dev:nixos.org) room on Matrix to see if anyone is working on a fix. I thought someone was, but I don’t recall who offhand and don’t think there is an issue or draft PR tracking that work.
 
> Also, if these commits are meant to just all be reverted soon, it might make it easier for us if all of the changes in this PR were included in a single commit.

Depending on whether the preference is to do a localized fix here or fix it Haskell-wide at the cost of rebuilds, I can squash or push that alternative instead.

reckenrode commented:
Pinging @toonn for awareness since he’s working on the cc-wrapper changes.

@cdepillabout I went ahead and squashed the commits as requested.

I skimmed the list of problematic dependencies in the Nix Tools report for the latest staging-next evaluation, but the other Haskell issues don’t appear to be related to this one. It’s not worth  the mass rebuilds to fix these packages at a lower level.

cdepillabout commented:
> My primary concern with modifying haskell-modules/generic-builder.nix until cc-wrapper is fixed is the number of rebuilds it will cause.
>
> The other Haskell issues don’t appear to be related to this one. It doesn’t seem not worth the mass rebuilds to use a more general fix.

The thing about the Haskell stuff in Nixpkgs is that it is commonly used by end users, to build their own packages not included in Nixpkgs.

If there is a way we can fix it generally, so that individual packages/users don't have to worry about it, I'd say we should do it.

My personal opinion is that it is more important to not break end-users' code, rather than trying to minimalize rebuilds on Hydra.

reckenrode commented:
Sure, I can do that. Thank you for clarifying. I’ve been erring on the side of avoiding rebuilds if possible, but I can put the fix in `haskell-modules/generic-builder.nix` so no one’s code breaks since that is an understandably higher priority.

(I assume that pushing the fix to staging instead is undesirable because there would be a risk of end-user code breaking before the next staging makes it way to master.)

I’ve tried to avoid rebuilds on Linux with my change. It’s narrowly targeted at clang stdenvs and only those using libc++, which Linux does not use by default. Once I confirm Cachix builds, I’ll update this PR.

reckenrode committed:
haskell-modules/generic-builder.nix: work around libc++abi issue

Work around clang’s not linking libc++abi when invoked as `clang` or
`cc` even when it’s needed to link a C++ library or C++ code.

See https://github.com/NixOS/nixpkgs/issues/166205.

reckenrode commented:
Cachix successfully built on both aarch64-darwin and x86_64-darwin with just this patch.

cdepillabout commented:
@reckenrode Thanks!

This seems reasonable.  Maybe @sternenseemann could do a final check, and we could merge in if he signs off on this.

amjoseph-nixpkgs commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -811,5 +811,12 @@ stdenv.mkDerivation ({
+
+# Ensure libc++abi is linked even when clang is invoked as just `clang` or `cc`.
+# Works around https://github.com/NixOS/nixpkgs/issues/166205.
+# This can be dropped once a fix has been committed to cc-wrapper.
```
Has a PR with the proposed fix to `cc-wrapper` been opened?

reckenrode replied:
@toonn is has been working on a fix. I don’t believe he’s opened a PR for it yet.

amjoseph-nixpkgs commented:
The Hydra log from the top of this issue has been deleted so I can't see the failure.

I suspect the root cause is that Darwin's `libcxx` doesn't have `isLLVM=true` even though it is LLVM-based:

https://github.com/NixOS/nixpkgs/blob/ab5d8762780613cc7f99132ba29f8e6ef950178e/pkgs/build-support/cc-wrapper/default.nix#L528-L533

reckenrode commented:
> I suspect the root cause is that Darwin's `libcxx` doesn't have `isLLVM=true` even though it is LLVM-based:

As I understand it, Darwin is not `isLLVM` because it doesn’t use enough of the toolchain. Historically, it used clang and libc++. That was it. This release, it’s also using some of the tools that are drop-in replacements, but not all of them are. In particular, it’s not using lld because lld isn’t fully capable of replacing ld64. I’d like to explore using lld 18 at least to build the stdenv for 24.11, but I’m still not sure whether it will be compatible enough to be the default linker on Darwin.

> https://github.com/NixOS/nixpkgs/blob/ab5d8762780613cc7f99132ba29f8e6ef950178e/pkgs/build-support/cc-wrapper/default.nix#L528-L533

Darwin’s `nix-support/libcxx-flags` already includes both `-stdlib=libc++` and `-lc++abi`. It can build and link C++ code fine when it is invoked as `clang++`, but it doesn’t include the C++ stuff when it is invoked as just `clang`. Unfortunately, unwrapped clang *does*, and some packages (such as Haskell and Bazel) expect to invoke `clang`  even when building and linking to C++.

For example, given (and pardon my C++98):

    #include <iostream>
    
    int main() {
        std::cout << "Hello, GitHub!" << std::endl;
    }

Compiling it with Xcode’s clang: `clang -o test test.cc -lc++` results in a working binary. Doing the same with the wrapped clang from nixpkgs results in:

    test.cc:1:10: fatal error: 'iostream' file not found
    #include <iostream>
             ^~~~~~~~~~
    1 error generated.

Which is because it’s ignoring the contents of `nix-support/libcxx-cxxflags`. If I specify the contents of that file on the command-line (or invoke it as `clang++`, it fails like this (and this is the kind of error affecting Haskell and Bazel):

```
Undefined symbols for architecture arm64:
  "std::terminate()", referenced from:
      ___clang_call_terminate in test-9e5058.o
  "___cxa_begin_catch", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in test-9e5058.o
      ___clang_call_terminate in test-9e5058.o
  "___cxa_end_catch", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in test-9e5058.o
  "___gxx_personality_v0", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char>>& std::__1::__put_character_sequence[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::basic_ostream<char, std::__1::char_traits<char>>&, char const*, unsigned long) in test-9e5058.o
      std::__1::char_traits<char>::length(char const*) in test-9e5058.o
      std::__1::ostreambuf_iterator<char, std::__1::char_traits<char>> std::__1::__pad_and_output[abi:v160006]<char, std::__1::char_traits<char>>(std::__1::ostreambuf_iterator<char, std::__1::char_traits<char>>, char const*, char const*, char const*, std::__1::ios_base&, char) in test-9e5058.o
      std::__1::ostreambuf_iterator<char, std::__1::char_traits<char>>::ostreambuf_iterator[abi:v160006](std::__1::basic_ostream<char, std::__1::char_traits<char>>&) in test-9e5058.o
      std::__1::basic_ios<char, std::__1::char_traits<char>>::widen[abi:v160006](char) const in test-9e5058.o
ld: symbol(s) not found for architecture arm64
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```

If I add `-lc++abi`, it finally builds and links.

vcunat commented:
FYI, binaries are being prepared already, as it seemed likely that this would go ahead soon:
https://hydra.nixos.org/eval/1801549

sternenseemann reviewed (commented):
Seems okay *if* it stays temporary. It's kind of funny how often something relating to C++ FFI breaks in Haskell exclusively on darwin…

vcunat commented:
OK, merged like this right now.  I suspect this hack will stay in 23.11, but I don't know if that would make it worth to improve the hack during some other rebuild.

reckenrode commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -811,5 +811,12 @@ stdenv.mkDerivation ({
+# Works around https://github.com/NixOS/nixpkgs/issues/166205.
+# This can be dropped once a fix has been committed to cc-wrapper.
+// lib.optionalAttrs (stdenv.cc.isClang && stdenv.cc.libcxx != null) {
+  env.NIX_LDFLAGS = "-l${stdenv.cc.libcxx.cxxabi.libName}";
```
Good catch. I pushed an update that preserves `env` while also avoiding any further rebuilds (since the staging-next job on Hydra is currently pointed to a branch with this commit cherry-picked).

reckenrode replied:
I opened #266878 to preserve `env` since this got merged before I could push the update.
